crates/flipperzero/src/subghz/device.rs

```diff
@@ -0,0 +1,193 @@
+        // function that takes a const pointer, so I don't believe that the data in the pointer get mutated.
+        unsafe {
+            subghz_devices_load_preset(self.device, furi_preset, preset_data);
+        }
+    }
+
+    // If the returned frequency from the internal API function call is 0, this function returns an error.
+    pub fn set_frequency(&mut self, freq: Frequency) -> Result<Frequency, SubGhzError> {
+        let freq_u32 = freq.value;
+        let actual_freq = unsafe { subghz_devices_set_frequency(self.device, freq_u32) };
+
+        if actual_freq != 0 {
+            Ok(Frequency::new::<hertz>(actual_freq))
+        } else {
+            Err(SubGhzError::UnableToSetFrequency)
+        }
+    }
+
+    pub fn is_frequency_valid(&mut self, freq: Frequency) -> bool {
+        unsafe { subghz_devices_is_frequency_valid(self.device, freq.value) }
+    }
+
+    pub fn set_async_mirror_pin(&mut self) {
+        unimplemented!("Should probably create safe abstraction around GpioPin first, have not investigated if this is needed");
+    }
+
+    pub fn get_data_gpio(&mut self) {
+        unimplemented!("Should probably create safe abstraction around GpioPin first, have not investigated if this is needed");
+    }
+
+    pub fn set_tx(&mut self) -> Result<(), SubGhzError> {
+        let result = unsafe { subghz_devices_set_tx(self.device) };
+
+        // Returns true if Ok
+        if result {
+            Ok(())
+        } else {
+            Err(SubGhzError::CannotTxOnFrequency)
+        }
+    }
+
+    pub fn flush_tx(&mut self) {
+        unsafe { subghz_devices_flush_tx(self.device) };
+    }
+
+    pub fn set_rx(&mut self) {
+        unsafe {
+            subghz_devices_set_rx(self.device);
+        }
+    }
+
+    pub fn flush_rx(&mut self) {
+        unsafe {
+            subghz_devices_flush_rx(self.device);
+        }
+    }
+
+    pub fn get_rssi(&mut self) -> Option<f32> {
+        let rssi_function = unsafe { self.device.interconnect.as_ref() }?.get_rssi?;
+        Some(unsafe { rssi_function() })
+    }
+
+    pub fn get_lqi(&mut self) -> Option<u8> {
+        let lqi_function = unsafe { self.device.interconnect.as_ref()?.get_lqi? };
+        Some(unsafe { lqi_function() })
+    }
+
+    pub fn rx_pipe_not_empty(&mut self) -> bool {
+        unsafe { subghz_devices_rx_pipe_not_empty(self.device) }
+    }
+
+    pub fn is_rx_crc_valid(&mut self) -> bool {
+        unsafe { subghz_devices_is_rx_data_crc_valid(self.device) }
+    }
+
+    pub fn read_packet(&mut self, buf: &mut [u8]) -> usize {
+        let mut size = if buf.len() > 255 {
+            255_u8
+        } else {
+            buf.len() as u8
+        };
+
+        unsafe {
+            subghz_devices_read_packet(self.device, buf.as_mut_ptr(), &mut size);
+        };
+        size as usize
+    }
+
+    pub fn write_packet(&mut self, buf: &[u8]) -> Result<(), SubGhzError> {
+        if buf.len() > 255 {
+            return Err(SubGhzError::PacketTooLong);
```

Comment:
I suggest adding `len` field to this error variant

Reply:
Good idea. I should also probably do this with the other variants as well. For now I will just do it for the PacketTooLong variant